File: src/main/java/io/github/dqualizer/dqlang/types/dam/architecture/RESTEndpoint.java
package io.github.dqualizer.dqlang.types.dam.architecture;

import io.github.dqualizer.dqlang.types.dam.Identifiable;
import kotlin.text.Regex;
import lombok.Data;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.http.HttpMethod;

import java.util.*;

@Data
@Document
public class RESTEndpoint extends ArchitectureEntity {

    public enum EndpointComponentType {
        PathVariable,
        QueryParameter,
        RequestBody,
        Header
    }

    @Data
    public static class EndpointComponent {
        private EndpointComponentType type;
        private String format;

        public final boolean checkFormat(String input) {
            Regex regex = new Regex(format);
            return regex.matches(input);
        }

    }

    private String name;
    private String route;

    private Map<EndpointComponentType, Set<EndpointComponent>> components = new HashMap<>();

    private Set<HttpMethod> methods = new HashSet<>();

    private ResponseDescription responseDescription;

    public Set<EndpointComponent> getComponentsOfType(EndpointComponentType type) {
        return components.getOrDefault(type, Collections.emptySet());
    }

}

File: src/main/kotlin/io/github/dqualizer/dqlang/types/dam/architecture/RESTEndpoint.kt
package io.github.dqualizer.dqlang.types.dam.architecture

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonProperty

data class RESTEndpoint @JsonCreator constructor(
  val name: String,
  // reference to a code component, for example a method
  @JsonProperty("code_component")
  val codeComponent: String,
  val route: String,
  val parameter: Set<EndpointParameter> = HashSet(),
  val methods: Set<EndpointMethod> = HashSet(),
  @JsonProperty("response_description")
  val responseDescription: ResponseDescription? = null
) {
  enum class EndpointParameterType {
    PathVariable,
    QueryParameter,
    RequestBody,
    Header
  }

  /**
   * We use our own enum since the Spring HttpMethod is not working with RabbitMQ serialization
   */
  enum class EndpointMethod {
    GET,
    HEAD,
    POST,
    PUT,
    PATCH,
    DELETE,
    OPTIONS,
    TRACE
  }

  data class EndpointParameter(
    val type: EndpointParameterType? = null,
    val data: String? = null
  ) {
    fun checkFormat(input: String?): Boolean {
      val regex = Regex(data!!)
      return regex.matches(input!!)
    }
  }

  fun getParameterOfType(type: EndpointParameterType): Set<EndpointParameter> {
    return parameter.filter { type == it.type }.toHashSet()
  }
}
